File: src/util/polish.ts
// On-device transcript cleanup, no model, no network. Wispr-style "Flow" formatting.

const FILLERS = /\b(um+|uh+|erm+|uhh+|hmm+|mm+hmm|y'?know)\b/gi;

// Spoken punctuation / layout commands → real symbols. Order matters (longest first).
const COMMANDS: [RegExp, string][] = [
  [/\b(new paragraph|next paragraph)\b/gi, '\n\n'],
  [/\b(new line|next line)\b/gi, '\n'],
  [/\b(bullet point|new bullet|bullet)\b/gi, '\n• '],
  [/\b(numbered list|number list)\b/gi, '\n1. '],
  [/\b(open paren|open parenthesis)\b/gi, '('],
  [/\b(close paren|close parenthesis)\b/gi, ')'],
  [/\b(open quote|quote)\b/gi, '“'],
  [/\b(close quote|unquote|end quote)\b/gi, '”'],
  [/\b(exclamation mark|exclamation point)\b/gi, '!'],
  [/\b(question mark)\b/gi, '?'],
  [/\b(full stop|period)\b/gi, '.'],
  [/\b(comma)\b/gi, ','],
  [/\b(colon)\b/gi, ':'],
  [/\b(semicolon)\b/gi, ';'],
  [/\b(hyphen|dash)\b/gi, '-'],
  [/\b(smiley face|smiley)\b/gi, ':)'],
];

function applyCommands(text: string): string {
  let t = text;
  for (const [re, sym] of COMMANDS) t = t.replace(re, sym);
  return t;
}

// Spoken EDITING commands. Destructive, so only applied to live dictation, never
// to imported recordings (a call participant saying "scratch that" must not
// delete text). No model, on-device.
const SCRATCH = /[^.!?\n]*\b(?:scratch|strike|delete|cross)\s+that\b[.,!?]*/gi;
const LAST_WORD = /(?:\S+)\s+\b(?:scratch|delete|cross)\s+(?:the\s+)?last word\b[.,!?]*/gi;
const LAST_LINE = /\b(?:scratch|delete|cross)\s+(?:this |the |that )?last line\b[.,!?]*/i;

export function applyVoiceCommands(text: string): string {
  let t = text.replace(SCRATCH, '').replace(LAST_WORD, '');
  if (LAST_LINE.test(t)) {
    const out: string[] = [];
    for (const line of t.split('\n')) {
      const m = line.match(LAST_LINE);
      if (m) {
        const before = line.slice(0, m.index).trim();
        if (!before && out.length) out.pop();
        continue;
      }
      out.push(line);
    }
    t = out.join('\n');
  }
  return t;
}

function tidy(text: string): string {
  let t = text
    .replace(/[ \t]+([,.!?;:])/g, '$1') // no space before punctuation
    .replace(/([,.!?;:])(?=[^\s\n])/g, '$1 ') // ensure space after punctuation
    .replace(/[ \t]{2,}/g, ' ') // collapse spaces
    .replace(/[ \t]*\n[ \t]*/g, '\n') // trim around newlines
    .replace(/\n{3,}/g, '\n\n')
    .trim();

  // Capitalize sentence starts and line starts.
  t = t.replace(/(^|[.!?]\s+|\n)([a-z])/g, (_m, pre, ch) => pre + ch.toUpperCase());
  // Standalone "i" -> "I"
  t = t.replace(/\bi\b/g, 'I');
  return t;
}

export function polish(text: string): string {
  if (!text.trim()) return text;
  return tidy(applyCommands(text).replace(FILLERS, ' '));
}

File: windows/Scribe/Polish.cs
using System.Text.RegularExpressions;

namespace Scribe;

static class Polish
{
  const RegexOptions I = RegexOptions.IgnoreCase | RegexOptions.CultureInvariant;

  static readonly Regex Fillers =
    new(@"\b(um+|uh+|erm+|uhh+|hmm+|mm+hmm|y'?know)\b", I);

  static readonly (Regex Re, string Sym)[] Commands =
  {
    (new(@"\b(new paragraph|next paragraph)\b", I), "\n\n"),
    (new(@"\b(new line|next line)\b", I), "\n"),
    (new(@"\b(bullet point|new bullet|bullet)\b", I), "\n• "),
    (new(@"\b(numbered list|number list)\b", I), "\n1. "),
    (new(@"\b(open paren|open parenthesis)\b", I), "("),
    (new(@"\b(close paren|close parenthesis)\b", I), ")"),
    (new(@"\b(open quote|quote)\b", I), "“"),
    (new(@"\b(close quote|unquote|end quote)\b", I), "”"),
    (new(@"\b(exclamation mark|exclamation point)\b", I), "!"),
    (new(@"\b(question mark)\b", I), "?"),
    (new(@"\b(full stop|period)\b", I), "."),
    (new(@"\b(comma)\b", I), ","),
    (new(@"\b(colon)\b", I), ":"),
    (new(@"\b(semicolon)\b", I), ";"),
    (new(@"\b(hyphen|dash)\b", I), "-"),
    (new(@"\b(smiley face|smiley)\b", I), ":)"),
  };

  static string ApplyCommands(string text)
  {
    foreach (var (re, sym) in Commands) text = re.Replace(text, sym);
    return text;
  }

  static readonly Regex Scratch =
    new(@"[^.!?\n]*\b(?:scratch|strike|delete|cross)\s+that\b[.,!?]*", I);
  static readonly Regex LastWord =
    new(@"(?:\S+)\s+\b(?:scratch|delete|cross)\s+(?:the\s+)?last word\b[.,!?]*", I);
  static readonly Regex LastLine =
    new(@"\b(?:scratch|delete|cross)\s+(?:this |the |that )?last line\b[.,!?]*", I);

  // Destructive: apply to live dictation only, never imported recordings.
  public static string ApplyVoiceCommands(string text)
  {
    var t = LastWord.Replace(Scratch.Replace(text, ""), "");
    if (LastLine.IsMatch(t))
    {
      var outLines = new List<string>();
      foreach (var line in t.Split('\n'))
      {
        var m = LastLine.Match(line);
        if (m.Success)
        {
          var before = line[..m.Index].Trim();
          if (before.Length == 0 && outLines.Count > 0) outLines.RemoveAt(outLines.Count - 1);
          continue;
        }
        outLines.Add(line);
      }
      t = string.Join("\n", outLines);
    }
    return t;
  }

  static readonly Regex NoSpaceBeforePunct = new(@"[ \t]+([,.!?;:])");
  static readonly Regex SpaceAfterPunct = new(@"([,.!?;:])(?=[^\s\n])");
  static readonly Regex CollapseSpaces = new(@"[ \t]{2,}");
  static readonly Regex TrimNewlines = new(@"[ \t]*\n[ \t]*");
  static readonly Regex TooManyNewlines = new(@"\n{3,}");
  static readonly Regex SentenceStart = new(@"(^|[.!?]\s+|\n)([a-z])");
  static readonly Regex StandaloneI = new(@"\bi\b");

  static string Tidy(string text)
  {
    var t = NoSpaceBeforePunct.Replace(text, "$1");
    t = SpaceAfterPunct.Replace(t, "$1 ");
    t = CollapseSpaces.Replace(t, " ");
    t = TrimNewlines.Replace(t, "\n");
    t = TooManyNewlines.Replace(t, "\n\n");
    t = t.Trim();
    t = SentenceStart.Replace(t, m => m.Groups[1].Value + char.ToUpperInvariant(m.Groups[2].Value[0]));
    t = StandaloneI.Replace(t, "I");
    return t;
  }

  public static string Format(string text)
  {
    if (string.IsNullOrWhiteSpace(text)) return text;
    return Tidy(Fillers.Replace(ApplyCommands(text), " "));
  }
}
